[bc-breaking] reference option for conv produce a pattern instead of reference conv module

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62215
* __->__ #61942

Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

jerryzh168 committed:
[bc-breaking] reference option for conv produce a pattern instead of reference conv module

Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

vkuzo commented on torch/nn/intrinsic/qat/modules/conv_fused.py:
```diff
@@ -303,6 +303,32 @@ def forward(self, input):
     def from_float(cls, mod):
         return super(ConvBnReLU1d, cls).from_float(mod)
 
+    def to_float(self):
```
would it be possible to reuse code between 1d/2d/3d same as `from_float` is doing? Or is there some blocker

jerryzh168 replied:
probably not, let me try if I can combine them

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

vkuzo commented on torch/nn/intrinsic/qat/modules/conv_fused.py:
```diff
@@ -344,6 +388,24 @@ def __init__(self,
                            padding, dilation, False, _pair(0), groups, bias, padding_mode,
                            eps, momentum, freeze_bn, qconfig, dim=2)
 
+    def to_float(self):
```
is there a way for this to call `_ConvBnNd.to_float` to reuse more code?

jerryzh168 replied:
oh sorry, I have implemented this in ConvBnNd, forgot to delete the extra code

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -526,25 +526,83 @@ def convert(self,
+                # to get scale and zero_point for weight
+                if isinstance(
+                        self.conv,
+                        (torch.nn.qat.Conv2d,
```
nit, could this be a mapping specified on top of the file and used as a var here?  Just to reduce LOC in this function

jerryzh168 replied:
sure

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -526,25 +526,83 @@ def convert(self,
+                         torch.nn.intrinsic.qat.ConvBn3d,
+                         torch.nn.intrinsic.qat.ConvBnReLU3d,
+                         torch.nn.intrinsic.qat.ConvReLU3d)):
+                    # case 1. converting qat conv module to
```
thanks for the comments!

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

jerryzh168 committed:
Update on "[bc-breaking] reference option for conv produce a pattern instead of reference conv module"


Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).
